src/main/java/io/cryostat/discovery/KubeApiDiscovery.java

```diff
@@ -0,0 +1,623 @@
+                    // of how the namespaces within the discovery tree are mapped.
+                    var result = new HashMap<String, SharedIndexInformer<Endpoints>>();
+                    kubeConfig
+                            .getWatchNamespaces()
+                            .forEach(
+                                    ns -> {
+                                        result.put(
+                                                ns,
+                                                client().endpoints()
+                                                        .inNamespace(ns)
+                                                        .inform(
+                                                                new EndpointsHandler(),
+                                                                ENDPOINTS_INFORMER_RESYNC_PERIOD));
+                                        logger.infov(
+                                                "Started Endpoints SharedInformer for"
+                                                        + " namespace \"{0}\"",
+                                                ns);
+                                    });
+                    return result;
+                }
+            };
+
+    @Transactional
+    void onStart(@Observes @Priority(1) StartupEvent evt) {
+        if (!(enabled())) {
+            return;
+        }
+
+        if (!available()) {
+            logger.errorv("{0} enabled but environment is not Kubernetes!", getClass().getName());
+            return;
+        }
+
+        DiscoveryNode universe = DiscoveryNode.getUniverse();
+        if (DiscoveryNode.getRealm(REALM).isEmpty()) {
+            DiscoveryPlugin plugin = new DiscoveryPlugin();
+            DiscoveryNode node = DiscoveryNode.environment(REALM, BaseNodeType.REALM);
+            plugin.realm = node;
+            plugin.builtin = true;
+            universe.children.add(node);
+            node.parent = universe;
+            plugin.persist();
+            universe.persist();
+        }
+
+        logger.infov("Starting {0} client", REALM);
+    }
+
+    @Transactional
+    void onAfterStart(@Observes StartupEvent evt) {
+        safeGetInformers();
+    }
+
+    void onStop(@Observes ShutdownEvent evt) {
+        if (!(enabled() && available())) {
+            return;
+        }
+
+        logger.infov("Shutting down {0} client", REALM);
+        safeGetInformers()
+                .forEach(
+                        (ns, informer) -> {
+                            informer.close();
+                            logger.infov(
+                                    "Closed Endpoints SharedInformer for namespace \"{0}\"", ns);
+                        });
+    }
+
+    boolean enabled() {
+        return enabled;
+    }
+
+    boolean available() {
+        try {
+            boolean hasNamespace = StringUtils.isNotBlank(kubeConfig.getOwnNamespace());
+            return kubeConfig.kubeApiAvailable() || hasNamespace;
```

Comment:
Should this be `&&`?

Reply:
I actually ported this from `2.x`. Though, it might be beneficial for environment such as local `kubeConfig` where the fabric can pick up info from there? Otherwise, I think `&&` sounds good.

Reply:
The local `kubeConfig` idea is interesting, but I guess there would be other things to implement to support it, and I don't think anything like that is planned right now.

Reply:
Sounds good I will update it to `&&` for now :D